### ISSUE TITLE
DoubleQuantilesSketch

### ISSUE THREAD

**Issue by churromorales** [apache/datasketches-cpp#263]
Is there any interest in providing the DoubleQuantilesSketch in cpp?  The java library contains it and druid only supports the DoubleQuantilesSketch at the moment.  We would like to send the serialized sketch to druid and it would be really useful to have this sketch in cpp.  Thank you.

**Comment by AlexanderSaydakov** [comment]
Yes, we are working on it.
https://github.com/apache/datasketches-cpp/commits/quantiles

**Comment by churromorales** [comment]
Awesome @AlexanderSaydakov  Thank you.

**Comment by AlexanderSaydakov** [comment]
Sorry, we don't have a particular timeline.

**Comment by the1schwartz** [comment]
@AlexanderSaydakov do you have an update on progress on getting support for DoubleQuantilesSketch?

**Comment by jmalkin** [comment]
It is pretty clearly insufficiently tested. I strongly suggest against trying to use it in its current form if you want things to work properly.

**Comment by jmalkin** [comment]
Well, I think I did at least fix serialization vs java.

**Comment by the1schwartz** [comment]
@jmalkin I am mostly interested in gettting serialization to work. Do you have a fix for this?

**Comment by the1schwartz** [comment]
Ah I can see your commit in repo. Thanks @jmalkin

**Comment by the1schwartz** [comment]
Hmm I get this message: `Error: Failed to sample data: Possible corruption: Must be v2, empty, or compact and ordered. Flags field: 1000, SerVer: 3` when trying to ingest the datasketch serialized using the datasketches-cpp (quantiles branch) into Druid

**Comment by the1schwartz** [comment]
I think the serialize function might be missing an "ordered" parameter to which dictate if the base_buffer should be sorted or not before serializing the datasketch

**Comment by the1schwartz** [comment]
I have been testing (locally) with the latest version of Druid. I think it is using the datasketches java library in its quantile datasketches extension as I got the same error when I tried to deserialize the data serialized using the quantile datasketches c++ library.

**Comment by jmalkin** [comment]
Ah-ha.. I think I misread the error message. It's an OR of those conditions (with "compact and ordered" as one). Ugh, that means there's a side-effect of sorting on a `const` method. Was hoping to avoid that :(

**Comment by the1schwartz** [comment]
Yeah I also had to read it twice to fully understand the message first I read it.

**Comment by jmalkin** [comment]
I still don't have any large-scale characterization tests to ensure things work, but I was at least able to load serialized c++ sketches directly in java just now. Merging and a python wrapper will happen before I do characterization -- the goal is to use the latter for non-speed characterization tests.

**Comment by jmalkin** [comment]
I owe an update here. Been busy with other stuff and fighting to make time for this. I think the sketch in the branch is working as expected. I dunno how performant it is, but I'm running an accuracy test now and so far it seems to be matching the old data we have on the website for the java version.

I'm a bit less confident that merging is always correct. I mean I certainly think it is or I wouldn't have started testing things, but I don't have a standard test suite with mixing sketches with different values of k, for instance.

The upshot is that I am (finally) confident that this is almost ready to be merged. And I think we'll try to cut a release very soon thereafter, although that'll probably involve reconciling some minor conflicts.

**Comment by jmalkin** [comment]
Did enough testing to submit PR #271
